pylint/checkers/variables.py

```diff
@@ -2215,10 +2239,7 @@ def _is_variable_violation(
                 )
             # check if we have a nonlocal
             elif node.name in defframe.locals:
-                maybe_before_assign = not any(
-                    isinstance(child, nodes.Nonlocal) and node.name in child.names
-                    for child in defframe.get_children()
-                )
+                maybe_before_assign = not _is_nonlocal_name(node, defframe)
```

Comment:
Hm. Is this the part of your change responsible for that new false negative? Or something else?

Reply:
not quite, it is related to `has_nonlocal_binding()`. it does a search for nonlocals across all enclosing frames and silences the message when consumer filters out all nodes

Reply:
Ah. So in the `while frame:` loop, can we stop the search after reaching an enclosing function? That would seem to solve the issue of the missing `nonlocal callback` in `inner()`.

Reply:
that makes sense, what you're saying is we search up to the frame that encloses both the usage and assignment? in that case we need to look at both the "name" node and "found" node i guess

Reply:
Note similar logic in [this function](https://github.com/pylint-dev/pylint/blob/main/pylint/checkers/base/basic_error_checker.py#L393-L413) which could perhaps be refactored (into checkers/utils.py?) to serve both situations.

Reply:
thanks guys, let's hold off on merging this one. i'll make the updates later

Reply:
resolving this thread, continuing with this at #10075